opentimelineio/adapters/fcp_xml.py

```diff
@@ -562,95 +1375,129 @@ def _build_item_timings(
 def _build_file(media_reference, br_map):
-    file_e = cElementTree.Element('file')
+    file_e = _element_with_item_metadata("file", media_reference)
 
     available_range = media_reference.available_range
     url_path = _url_to_path(media_reference.target_url)
 
-    _insert_new_sub_element(file_e, 'name', text=os.path.basename(url_path))
-    file_e.append(_build_rate(available_range.start_time))
-    _insert_new_sub_element(
+    file_name = (
+        media_reference.name if media_reference.name
+        else os.path.basename(url_path)
+    )
+    _append_new_sub_element(file_e, 'name', text=file_name)
+    _append_new_sub_element(file_e, 'pathurl', text=media_reference.target_url)
+
+    # timing info
+    file_e.append(_build_rate(available_range.start_time.rate))
+    _append_new_sub_element(
         file_e, 'duration',
         text='{:.0f}'.format(available_range.duration.value)
     )
-    _insert_new_sub_element(file_e, 'pathurl', text=media_reference.target_url)
 
     # timecode
-    timecode = available_range.start_time
-    timecode_e = _insert_new_sub_element(file_e, 'timecode')
-    timecode_e.append(_build_rate(timecode))
-    # TODO: This assumes the rate on the start_time is the framerate
-    _insert_new_sub_element(
-        timecode_e,
-        'string',
-        text=otio.opentime.to_timecode(timecode, rate=timecode.rate)
-    )
-    _insert_new_sub_element(
-        timecode_e,
-        'frame',
-        text='{:.0f}'.format(timecode.value)
+    ref_tc_metadata = media_reference.metadata.get(META_NAMESPACE, {}).get(
+        "timecode"
     )
-    display_format = (
-        'DF' if (
-            math.ceil(timecode.rate) == 30 and math.ceil(timecode.rate) != timecode.rate
-        ) else 'NDF'
+    tc_element = _build_timecode_from_metadata(
+        available_range.start_time, ref_tc_metadata
     )
-    _insert_new_sub_element(timecode_e, 'displayformat', text=display_format)
+    file_e.append(tc_element)
 
     # we need to flag the file reference with the content types, otherwise it
     # will not get recognized
-    file_media_e = _insert_new_sub_element(file_e, 'media')
-    content_types = []
-    if not os.path.splitext(url_path)[1].lower() in ('.wav', '.aac', '.mp3'):
-        content_types.append('video')
-    content_types.append('audio')
+    # TODO: We should use a better method for this. Perhaps pre-walk the
+    #       timeline and find all the track kinds this media is present in?
+    if not file_e.find("media"):
+        file_media_e = _get_or_create_subelement(file_e, "media")
 
-    for kind in content_types:
-        _insert_new_sub_element(file_media_e, kind)
+        audio_exts = {'.wav', '.aac', '.mp3', '.aif', '.aiff', '.m4a'}
+        has_video = (os.path.splitext(url_path)[1].lower() not in audio_exts)
+        if has_video and file_media_e.find("video") is None:
+            _append_new_sub_element(file_media_e, "video")
+
+        # All files have audio?
```

Comment:
I don't understand this comment

Reply:
In this pass I did some refactoring for clarity on the output side, but I was trying to minimally impact some of the structural decisions made. This comment was an attempt to call out that this piece of logic may not be sustainable. Will update with a better comment.